Userland/Libraries/LibAudio/MP3Loader.cpp

```diff
@@ -147,4 +180,679 @@ bool MP3LoaderPlugin::synchronize()
     return one_counter == 12;
 }
 
+Optional<MP3::MP3Frame> MP3LoaderPlugin::read_next_frame()
+{
+    while (!m_bitstream->has_any_error()) {
+        if (!synchronize()) {
+            m_bitstream->handle_any_error();
+            return {};
+        }
+
+        MP3::Header header = read_header();
+        if (header.id != 1 || header.sampling_frequency == -1 || header.bitrate == 0 || header.layer != 3) {
+            // Unsupported stream or misinterpreted garbage for // TODO: stream should jump back 2 bytes to not accidently jump over actual frame
+            dbgln("unsupported frame found");
+            continue;
+        }
+
+        return read_frame_data(header, false);
+    }
+    return {};
+}
+
+Optional<MP3::MP3Frame> MP3LoaderPlugin::read_frame_data(const MP3::Header& header, bool is_first_frame)
+{
+    MP3::MP3Frame frame = {};
+    frame.header = header;
+
+    read_side_information(frame);
+    if (m_bitstream->has_any_error())
+        return {};
+
+    auto maybe_buffer = ByteBuffer::create_uninitialized(header.slot_count);
+    if (!maybe_buffer.has_value())
+        return {};
+    auto& buffer = maybe_buffer.value();
+
+    size_t old_reservoir_size = m_bit_reservoir.size();
+    if (m_bitstream->read(buffer) != buffer.size())
+        return {};
+    if (m_bit_reservoir.write(buffer) != header.slot_count)
+        return {};
+
+    if (frame.main_data_begin > 0) {
+        if (is_first_frame)
+            return frame;
+        if (!m_bit_reservoir.discard_or_error(old_reservoir_size - frame.main_data_begin))
+            return {};
+    }
+
+    InputBitStream reservoir_stream(m_bit_reservoir);
+
+    for (size_t gr = 0; gr < 2; gr++) {
+        for (size_t ch = 0; ch < header.channel_count(); ch++) {
+            size_t scf_size = read_scale_factors(frame, reservoir_stream, gr, ch);
```

Comment:
These variables could use some longer, more descriptive names than "gr", "ch" and "scf" :P

Reply:
I kept it close to the standard. I can see though why that may look like a bunch of weirdness. Changed